src/api/uploads/MultiputPart.js

```diff
@@ -34,30 +51,264 @@ class MultiputPart extends Base {
+     * @param {MultiputConfig} config
+     * @param {Function} getPartsState
+     * @param {Function} getNumPartsUploading
+     * @param {Function} [onSuccess]
+     * @param {Function} [onProgress]
+     * @param {Function} [onError]
      * @return {void}
      */
-    constructor(options: Object, index: number, offset: number, size: number): void {
-        super(options);
+    constructor(
+        options: Object,
+        index: number,
+        offset: number,
+        size: number,
+        sessionId: string,
+        sessionEndpoints: Object,
+        config: MultiputConfig,
+        getPartsState: Function,
+        getNumPartsUploading: Function,
+        onSuccess?: Function,
+        onProgress?: Function,
+        onError?: Function
+    ): void {
+        super(options, sessionEndpoints, config);
 
         this.index = index;
         this.numDigestRetriesPerformed = 0;
         this.numUploadRetriesPerformed = 0;
         this.offset = offset;
-        this.sha256 = null;
         this.size = size;
         this.state = PART_STATE_NOT_STARTED;
         this.timing = {};
         this.uploadedBytes = 0;
+        this.data = {};
+        this.uploadUrl = `${this.uploadHost}/api/2.0/files/upload_sessions/${sessionId}`;
+        this.config = config;
+        this.rangeEnd = offset + size - 1;
+        this.onSuccess = onSuccess || noop;
+        this.onError = onError || noop;
+        this.onProgress = onProgress || noop;
+        this.getPartsState = getPartsState;
+        this.getNumPartsUploading = getNumPartsUploading;
     }
 
+    stringify = () =>
+        JSON.stringify({
+            index: this.index,
+            offset: this.offset,
+            size: this.size,
+            state: this.state,
+            uploadedBytes: this.uploadedBytes,
+            numUploadRetriesPerformed: this.numUploadRetriesPerformed,
+            numDigestRetriesPerformed: this.numDigestRetriesPerformed,
+            sha256: this.sha256,
+            xhr: this.xhr.xhr,
+            timing: this.timing
+        });
+
+    /**
+     * Returns file part associated with this Part.
+     *
+     * @return {Object}
+     */
+    getPart = (): Object => this.data.part;
+
+    /**
+     * Uploads this Part via the API. Will retry on network failures.
+     *
+     * @return {void}
+     */
+    upload = (): void => {
+        if (this.isDestroyed() || !this.sha256 || !this.blob) {
+            return;
+        }
+
+        const clientEventInfo = {
+            documentHidden: document.hidden,
+            digest_retries: this.numDigestRetriesPerformed,
+            timing: this.timing,
+            parts_uploading: this.getNumPartsUploading()
+        };
+
+        const headers = {
+            'Content-Type': 'application/octet-stream',
+            Digest: `sha-256=${this.sha256}`,
+            'Content-Range': `bytes ${this.offset}-${this.rangeEnd}/${this.size}`,
+            'X-Box-Client-Event-Info': JSON.stringify(clientEventInfo)
```

Comment:
Double check if this header is whitelisted in our headers for app.box.com

Reply:
Good point, it is not. @wenboyu2 you're going to need to to add this header to our CORS whitelist which I'll message you.

I always go back and forth about whether or not to remove this shortcut... This is gonna cause a site issue or at a minimum block some release someday.

Reply:
I'm not sure if whitelisting is necessary as this request hits UPXY directly (and UPXY also generates the CORS response)

Reply:
This is the CORS shortcut we have in nginx (since nginx responds a lot faster than PHP):

```
if ($http_origin ~* (^https://(.+\.)?(app|ent)\.(box\.com|boxcn\.net|boxenterprise\.net)$)) {
    set $cors "box";
}

if ($request_method = 'OPTIONS') {
    set $cors "${cors}options";
}

# If CORS pre-flight request has subdomain app or ent on the root domain box.com, boxcn.net,
# or boxenterprise.net, handle in nginx to improve performance
if ($cors = "boxoptions") {
    // Set CORS headers + list of allowed headers. The list of headers needs to be manually specified since not all browser support Access-Control-Allow-Headers: * yet
}
```

This conf is included in the api.box.com nginx template, so you're maybe right that it won't get triggered if we're hitting UPXY directly.